src/figuretype/trader.c

```diff
@@ -922,10 +903,6 @@ void figure_trade_ship_action(figure *f)
                     city_message_post(1, MESSAGE_NAVIGATION_IMPOSSIBLE, 0, 0);
                     city_message_increase_category_count(MESSAGE_CAT_BLOCKED_DOCK);
                 }
-            } else if (f->wait_ticks++ >= FIGURE_REROUTE_DESTINATION_TICKS) {
```

Comment:
This code had the advantage of allowing ships to constantly check for a newer free dock even when they are going for another one.

From what I gather, removing this code makes the ship still go to a busy/ further away dock even if one got free in the meantime. 

Granted the bugs should be fixed, but can't the "look for the best available dock" behaviour be kept?

Reply:
@crudelios Good question. Checking for better free docks is still happening during most of the ship's lifetime, only when the ship is going to dock queue `case FIGURE_ACTION_113_TRADE_SHIP_GOING_TO_DOCK_QUEUE:`, not going directly to the dock's work grid, which that line belongs to. The reason is now ships never go to dock directly, it must anchor in queue even if it's free. Always anchor-first. I'll explain below.

Vanilla C3 had ships go directly to dock from spawning point (river_entry) if dock is free. Ships anchoring in queue near the dock keeps waiting for a distant ship to arrive, not fair.

The solution I came up with was to never allow ships to go to dock directly. Instead they anchor near it in queue (even if it's free) for few ticks, then go directly to this dock to start the trade from the anchoring point normally. So if a ship is going to a dock (which means it's already anchoring very near) it shouldn't care about other docks because it's almost there.

That means when a dock had no available anchoring queue grids, like [in this case](https://github.com/Keriew/augustus/issues/240#issuecomment-786281933) then no ships at all can queue and then no ship will eventually enter the dock! That's why anchoring grid points were checked and adjusted in building/dock.c `building_dock_get_ship_request_tile`

The only exception is when a ship is already anchoring, and a _near by_ `MAX_DISTANCE_FOR_REROUTING = 50` tiles away dock becomes free, then it go directly to that dock work grid, preventing other ships from entering. Also in this case no need to check for better docks because it's already near a free dock.

Rerouting a ship that is going to a dock already is tricky because you will have to release the dock's lock in `dock->data.dock.trade_ship_id`
I'm open for other suggestions other than the anchor-first rule.